rfcs/0007-filtering.md

```diff
@@ -0,0 +1,241 @@
+
+This proposal will suggest adding server side filtering to API calls that return EOBs to the client in order to 
+reduce on the amount of data that needs to be transferred. 
+
+## Table of Contents
+[Table of Contents]: #table-of-contents
+
+* [RFC Proposal](#rfc-proposal)
+* [Table of Contents](#table-of-contents)
+* [Motivation](#motivation)
+* [Proposed Solution](#proposed-solution)
+    * [Proposed Solution: Detailed Design](#proposed-solution-detailed-design)
+    * [Proposed Solution: Unresolved Questions](#proposed-solution-unresolved-questions)
+    * [Proposed Solution: Drawbacks](#proposed-solution-drawbacks)
+    * [Proposed Solution: Notable Alternatives](#proposed-solution-notable-alternatives)
+* [Future Possibilities](#future-possibilities)
+
+## Motivation
+[Motivation]: #motivation
+
+A single BFD EOB retrieval call (UHC) with 2 million beneficiaries took 24 hours to retrieve three months of data.
+Larger PDPs can have in excess of 6 million beneficiaries.  To have a request take over 3 days is not acceptable and in
+fact the downloaded EOB files will have already been removed from S3.
+
+AB2D receives more data than needed from BFD when we make API calls to gather EOBs (Explanation of Benefits).  AB2D
+currently filters out irrelevant fields.  This might not be terrible for dozens of API calls, but for 
+millions of requests it can impact our performance significantly.  Since we only need a projection of the data
+(10 out of 69 top-level getters; 6 out of 42 fields in the ItemComponent), it makes
+sense for the server to send us only the data consumed.  This reduces bandwidth requirements for receiving the data.
+
+For example, GraphQL solves this problem by allowing clients to specify field selection (e.g. a database projection) at
+invocation time.
+
+## Proposed Solution
+[Proposed Solution]: #proposed-solution
+
+The solution adds code in the BFD server to configure which fields AB2D requires.  Non-required fields are nulled out
+before being transformed in the result.  The specific set of required fields should remain stable
+and not require modifications.
+
+The filtering will need to happen towards the end of processing. This is to avoid
+spaghetti code since transformations from the entity object to the FHIR object happen in several places throughout
+the codebase. 
+
+
+### Proposed Solution: Detailed Design
+[Proposed Solution: Detailed Design]: #proposed-solution-detailed-design
+
+The BFD_FIELD_SELECTION header enables this feature.  A value of **ALL** is current behavior, will be the default
+(if header is missing) and indicates to return all EOB fields.  A value of **AB2D_VERSION1** indicates return only
+the fields specified in this proposal.  By using an enumerated type instead of a boolean, this allows for different
+results to be specified in the future should the need arise and for different consumers to have specific field selections
+encoded.
```

Comment:
Please stick with idiomatic HTTP header naming conventions. See <https://stackoverflow.com/a/3561399> and elsewhere.

Reply:
X- prefix has been deprecated and its use is discouraged.

the official recommendation is to just name them sensibly without the "X-" prefix

There seems to be some variance in the community in that some folks still advocate using -X for application specific parameters (which this is).

Please suggest a header name that you like better.

Reply:
Sorry, I should have been more clear: I meant "please don't use underscores and follow the case conventions."